Feat/connect: Entropy check workflow in ResetDevice

szymonlesisz opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

implementation on [entropy check workflow](https://github.com/trezor/trezor-firmware/tree/andrewkozlik/display_random)

szymonlesisz commented on packages/connect/src/api/resetDevice.ts:
```diff
@@ -42,10 +48,70 @@ export default class ResetDevice extends AbstractMethod<'resetDevice', PROTO.Res
+            this.params.entropy_check = false;
+        }
+        // Entropy check workflow:
+        // https://github.com/trezor/trezor-firmware/blob/andrewkozlik/display_random/docs/common/message-workflows.md#entropy-check-workflow
```
TODO: update link

szymonlesisz commented:
i guess it was premature excitement regarding `@noble/hashes`.

it takes ~8sec. to perform `@noble/hashes/pbkdf2` in the web build :(

i'm fallbacking to WebCrypto available in node.js and web (takes ~300ms to calculate)

szymonlesisz commented on packages/connect/src/api/resetDevice.ts:
```diff
@@ -42,10 +48,71 @@ export default class ResetDevice extends AbstractMethod<'resetDevice', PROTO.Res
+            xpubs,
+        });
+        if (res.error) {
+            throw new Error(res.error);
```
TODO: return specific error to suite

szymonlesisz replied:
done

onvej-sl commented on packages/connect/src/api/resetDevice.ts:
```diff
@@ -42,10 +48,71 @@ export default class ResetDevice extends AbstractMethod<'resetDevice', PROTO.Res
-        return response.message;
+        if (this.params.entropy_check && this.device.unavailableCapabilities['entropyCheck']) {
+            // entropy check requested but not supported by the firmware
+            this.params.entropy_check = false;
```
From a security perspective, if the firmware does not support the entropy check, we should display the same error message as we would if the entropy check failed.

The reason is that a counterfeit trezor could bypass the entropy check by pretending not to support it, that is the trezor can present itself as having a lower firmware version than it actually does. However, we can instruct the user to update the firmware before generating their seed.

szymonlesisz replied:
this feature needs to be backward compatible with firmwares which does not support it for example afaik T1 will not be supported. 
i need to rely on something to make the decision whether `ResetDeviceContinue` will be sent or not.
following your example counterfeit trezor could bypass this check by pretending that its T1 model.

I think suite in the first onboarding step does FW version check (and asks you to update) then in the next step you create a seed.

We were discussing this version check with @andrewkozlik

onvej-sl commented on packages/connect/src/api/firmware/verifyEntropy.ts:
```diff
@@ -0,0 +1,167 @@
+    if (
+        type &&
+        [
+            BackupType.Slip39_Basic,
+            BackupType.Slip39_Advanced,
```
The implementation of slip39 in this file assumes that the backup is extendable, meaning that using a non-extendable backup will likely result in the "verifyEntropy xpub mismatch" error.

szymonlesisz replied:
i did it 1:1 with python-trezor implementation, which assumes that as well:
https://github.com/trezor/trezor-firmware/blob/ce803a5452c5d12eb7b026fa2c20b521957fb8c8/python/src/trezorlib/device.py#L268

```
seed = shamir_mnemonic.cipher.decrypt(
            secret, b"", iteration_exponent=1, identifier=0, extendable=True
        )
```

onvej-sl replied:
Fair enough.

However, I think both implementations should raise an exception in this case. Nonetheless, this is not a security issue.

onvej-sl commented on packages/connect/src/api/resetDevice.ts:
```diff
@@ -42,10 +48,71 @@ export default class ResetDevice extends AbstractMethod<'resetDevice', PROTO.Res
+        return currentData;
+    }
+
     async run() {
```
If the entropy check fails, a small error message is displayed to the user and the user can continue with the onboarding process. We should warn the user similarly to how they are warned when the authenticity check fails.

I think this has already been discussed in person so I'm adding this comment merely as a reminder.

szymonlesisz replied:
yes, i've left a TODO above: return specific error to suite

btw. when this happens user cannot continue with onboarding as the seed was not created (ResetDeviceFinish was not called)

onvej-sl commented on packages/connect/src/api/resetDevice.ts:
```diff
@@ -42,10 +48,71 @@ export default class ResetDevice extends AbstractMethod<'resetDevice', PROTO.Res
+    private async entropyCheck(prevData: EntropyRequestData): Promise<EntropyRequestData> {
+        const cmd = this.device.getCommands();
+        const paths = ["m/84'/0'/0'", "m/44'/60'/0'"];
+        const xpubs: Record<string, string> = {}; // <path, xpub>
```
We should verify that these XPUBs match the XPUBs later used by the suite to derive change and receive addresses.

I think this has already been discussed in person so I'm adding this comment merely as a reminder.

szymonlesisz replied:
yes, marked as "followup" above

szymonlesisz commented:
rebased, resolved conflicts (dependencies, yarn.lock) and implemented [recent changes](https://github.com/trezor/trezor-firmware/commit/28acbe48f02324e70ea4d7c91a3121af41b55de2) - renamed protobuf fields

szymonlesisz commented:
test `recovery/t1b1-dry-run.test.ts` and suite-web e2e are flaky? the are failing not only in this branch:
https://github.com/trezor/trezor-suite/actions/workflows/test-suite-desktop-e2e.yml
https://github.com/trezor/trezor-suite/actions/workflows/test-suite-web-e2e.yml

onvej-sl commented:
I am unable to "Resolve conversation" in my comments above. Feel free to resolve them if you can.

marekrjpolak commented on packages/connect/src/api/firmware/verifyEntropy.ts:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    // use bip39
+    return mnemonicToSeed(entropyToMnemonic(secret));
```
Have you considered using `@scure/bip39` instead of `bip39`? It's already in the repo because of some Ehtereum libs and also seems a bit better maintained, but it has to be given the wordlist.

```diff
- import { entropyToMnemonic, mnemonicToSeed } from 'bip39';
+ import { entropyToMnemonic, mnemonicToSeed } from '@scure/bip39';
+ import { bip39 } from '@trezor/crypto-utils';
...
- return mnemonicToSeed(entropyToMnemonic(secret));
+ return mnemonicToSeed(entropyToMnemonic(secret, [...bip39])).then(seed => Buffer.from(seed));
```

szymonlesisz replied:
i had to make `@trezor/crypto-utils` public because of this

komret commented on packages/connect/src/data/config.ts:
```diff
@@ -259,5 +259,9 @@ export const config = {
         },
+        {
+            capabilities: ['entropyCheck'],
+            min: { T1B1: '0', T2T1: '2.8.7', T2B1: '2.8.7' },
```
Shouldn't there be T3B1 and T3T1 as well?

szymonlesisz replied:
done

marekrjpolak commented on packages/connect/src/data/config.ts:
```diff
@@ -259,5 +259,9 @@ export const config = {
         },
+        {
+            capabilities: ['entropyCheck'],
+            min: { T1B1: '0', T2T1: '2.8.7', T2B1: '2.8.7' },
```
`T3T1` missing intentionally?

szymonlesisz replied:
not intentionally, i just forgot about them :)

szymonlesisz replied:
done

marekrjpolak commented on packages/connect/src/api/resetDevice.ts:
```diff
@@ -42,10 +48,75 @@ export default class ResetDevice extends AbstractMethod<'resetDevice', PROTO.Res
+
     async run() {
         const cmd = this.device.getCommands();
-        const response = await cmd.typedCall('ResetDevice', 'Success', this.params);
```
Different expected response from `ResetDevice` (`Success` before, `EntropyRequest` now) is because previously `ResetDevice` always returned `EntropyRequest`, which was acked from DeviceCommands which returned `Success`? 

I just want to make sure that there's no change in message flow apart from extending it.

andrewkozlik replied:
FYI, if it's any help the workflow is documented [here](https://github.com/trezor/trezor-firmware/blob/main/docs/common/message-workflows.md#resetdevice).

marekrjpolak replied:
Thanks, I know about this one. It's just that from our code it seemed that `ResetDevice` previously returned `Success`, but now I believe it was just because `EntropyRequest`/`EntropyAck` were hidden under the hood.

szymonlesisz replied:
well, not exactly
previous flow also returns EntropyRequest but it was handled by DeviceCommands._filterCommonTypes as we didnt work with EntropyRequest data until now. same flow is also used if `entropy_check` is set to false 

(Host) ResetDevice > (Trezor) EntropyRequest > (Host) EntropyAck > (Trezor) Success

marekrjpolak reviewed (commented):
Looks good, but I haven't tested it yet.

komret commented on packages/connect/src/api/firmware/verifyEntropy.ts:
```diff
@@ -0,0 +1,167 @@
+}: VerifyEntropyOptions) => {
+    try {
+        if (!trezorEntropy || !commitment || !strength || Object.keys(xpubs).length < 1) {
+            throw new Error('Missing verifyEntropy data');
```
When an error is thrown here, Suite resets to the default screen but device hangs on loading screen until the user interacts with it again via Suite.

![Screenshot 2025-01-03 at 16 52 37](https://github.com/user-attachments/assets/f63b4ac6-d9b4-4ff3-82fe-18306ed23464)

szymonlesisz replied:
fixed, `resetDevice` will send `Cancel` message before throwing `Failure_EntropyCheck`

bosomt commented:
**QA OK**

**Info:**
 - Suite version: desktop 25.1.1 (4403882efb2a4a802c3d77917d6350a4fae7663d)
 - Browser: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) TrezorSuite/25.1.1 Chrome/128.0.6613.186 Electron/32.2.6 Safari/537.36
 - OS: MacIntel
 - Screen: 1512x982
 - Device: Trezor T3B1 2.8.7 bitcoin-only (revision 8a254aa8eae82f99630df63f40e4d290066a3efc)
 - Transport: BridgeTransport 3.0.0-bundled.25.1.1
